design/one-pager-k8s-native-providers.md

```diff
@@ -0,0 +1,217 @@
+provider:
+
+![K8s Native Providers](./images/k8s-native-providers.png)
+
+## Technical Design
+
+### Claim Reconciler
+
+The claim controllers could use the same shared [claim reconciler] that other
+existing infrastructure stacks use currently, providing their own
+`ManagedConfigurators` to configure the managed resource instance using the
+referenced class.
+
+### Scheduler Reconciler
+
+Because the scheduling behavior for most Kubernetes-native infrastructure
+components will look very similar, a shared scheduler reconciler should be added
+to [crossplane-runtime]. It will look almost identical to the reconciler used
+for `KubernetesApplication` [scheduler controller]. In fact, the
+`KubernetesApplication` scheduler controller should likely make use of the
+shared `crossplane-runtime` scheduler reconciler after it is implemented.
+
+### Managed Reconciler
+
+Though this design is motivated by the desire to support Rook, the pattern can
+be generalized to be applicable to other Kubernetes-native infrastructure
+providers. While most existing infrastructure provider stacks must create a
+client wrapper library to use the CRUD functions of the provider API, every
+Kubernetes-native infrastructure stack will use the [client-go] SDK for CRUD
+operations. Therefore, the pluggable `Observe`, `Create`, `Update`, `Delete`
+functions in the [crossplane-runtime] can likely have a common implementation
+for Kubernetes-native infrastructure stacks.
+
+This can be done via generic `NativeConnecter` and `nativeExternal` types in
+`crossplane-runtime`. Similar to the `clusterConnecter` in the
+`KubernetesApplicationResource` controller, it would get the cluster connection
+information of the Kubernetes cluster that the managed resource was scheduled to
+by the scheduler controller, then return a `nativeExternal` type that implements
+the `ExternalClient` interface. Because external client functions will
+essentially be making sure the "external resource" in the target cluster matches
+the "managed resource" (i.e. they should be nearly identical Kubernetes
+resources), it should be possible to have generic implementations of the
+`ExternalClient` functions. However, if it proves difficult to do so, or it just
+does not work for a specific resource type, manual implementation of the
+`connecter` and `external` types would always be possible, just as they are for
+existing infrastructure stacks.
+
+Implementation of the shared managed reconciler for the managed CockroachDB
+`Cluster` controller (`cockroachdb.rook.crossplane.io/v1alpha1`) if a
+`nativeExternal` type was able to be utilized could look as follows:
+
+```go
+func (c *ClusterController) SetupWithManager(mgr ctrl.Manager) error {
+  r := resource.NewManagedReconciler(mgr,
+    resource.ManagedKind(v1alpha1.ClusterGroupVersionKind),
+    resource.WithExternalConnecter(&resource.NativeConnecter{client: mgr.GetClient(), kind: v1alpha1.Cluster}))
```

Comment:
Yeah, I could see many of these just providing their own implementation, which they are still open to do of course. However, in the cases where it is possible this might be nice. We could certainly just push this kind of functionality down the road and see if it is useful / desired after multiple k8s-native infrastructure providers have done it manually.